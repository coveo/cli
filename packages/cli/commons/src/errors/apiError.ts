import {red} from 'chalk';
import dedent from 'ts-dedent';
import {PrintableError, SeverityLevel} from './printableError';

export interface APIErrorResponse {
  message: string;
  errorCode: string;
  requestID: string;
}

export interface AxiosErrorFromAPI {
  response: {
    status: number;
    data: APIErrorResponse;
  };
}

export const APIErrorSchema = {
  type: 'object',
  required: ['errorCode', 'message'],
  properties: {
    errorCode: {
      type: 'string',
    },
    message: {
      type: 'string',
    },
  },
};

export const AxiosErrorFromAPISchema = {
  type: 'object',
  required: ['response'],
  properties: {
    response: {
      type: 'object',
      required: ['status', 'data'],
      properties: {
        status: {
          type: 'integer',
        },
        data: APIErrorSchema,
      },
    },
  },
};

export class APIError extends PrintableError {
  public constructor(
    error: APIErrorResponse | AxiosErrorFromAPI,
    tagLine?: string
  ) {
    super(SeverityLevel.Error);
    let errorCode: string;
    let message: string;
    let status: number | undefined;
    if (this.isFromAxios(error)) {
      errorCode = error.response.data.errorCode;
      message = error.response.data.message;
      status = error.response.status;
    } else {
      errorCode = error.errorCode;
      message = error.message;
    }

    this.name = `APIError - ${errorCode}`;
    this.message = dedent`
    ${tagLine}
    ${status ? `Status code: ${status}\n` : ''}
    Error code: ${red(errorCode)}
    Message: ${red(message)}
    `;
  }

  private isFromAxios(
    error: APIErrorResponse | AxiosErrorFromAPI
  ): error is AxiosErrorFromAPI {
    return 'response' in error;
  }
}
