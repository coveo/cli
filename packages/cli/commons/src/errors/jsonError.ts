import type {ValidationError} from 'jsonschema';

export function getPrettyJsonValidationErrors(
  errors: ValidationError[]
): string {
  const stackToPrettyError = (stack: string) =>
    ` - ${stack.replace(/^instance\./, '')}\n`;
  return errors
    .reduce<string>(
      (errors, error) => (errors += stackToPrettyError(error.stack)),
      '\n'
    )
    .replace(/\n$/, '');
}
