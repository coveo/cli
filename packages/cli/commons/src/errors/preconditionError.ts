import {PrintableError, SeverityLevel} from './printableError';

export interface PreconditionErrorOptions {
  category?: string;
  level?: SeverityLevel;
}

export class PreconditionError extends PrintableError {
  public constructor(
    public message: string,
    public options?: PreconditionErrorOptions
  ) {
    super(options?.level || SeverityLevel.Error);
    this.name = 'Precondition Error';
    if (this.options?.category) {
      this.name += ` - ${this.options?.category}`;
    }
  }
}
