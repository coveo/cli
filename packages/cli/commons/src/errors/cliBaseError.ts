import {CLIError} from '@oclif/core/lib/errors';
import {Chalk, red, yellow, cyan} from 'chalk';

export enum SeverityLevel {
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
}

export interface CLIBaseErrorInterface {
  level?: SeverityLevel;
  cause?: Error;
}

interface OClifCLIError extends Omit<CLIError, 'render' | 'bang' | 'oclif'> {}

export class CLIBaseError extends Error implements OClifCLIError {
  private static readonly defaultSeverity: SeverityLevel = SeverityLevel.Error;
  public name = 'CLI Error';

  public constructor(
    private error?: string | Error | CLIError,
    private options?: CLIBaseErrorInterface
  ) {
    super(error instanceof Error ? error.message : error, options);
  }

  public get stack(): string {
    return super.stack || '';
  }

  public get severityLevel(): SeverityLevel {
    return this.options?.level || CLIBaseError.defaultSeverity;
  }

  /**
   * Specific to internal oclif error handling
   */
  private get oclif() {
    return this.error instanceof CLIError ? this.error.oclif : {};
  }

  /**
   * Used by oclif to pretty print the error
   */
  private get bang() {
    let color: Chalk;

    switch (this.severityLevel) {
      case SeverityLevel.Info:
        color = cyan;
        break;

      case SeverityLevel.Warn:
        color = yellow;
        break;

      case SeverityLevel.Error:
      default:
        color = red;
        break;
    }
    return color('Â»');
  }
}
