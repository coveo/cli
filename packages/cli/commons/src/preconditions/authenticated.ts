import {PreconditionError} from '../errors/preconditionError';
import {
  AuthenticationStatus,
  getAuthenticationStatus,
} from '../platform/authenticatedClient';
import {Config} from '../config/config';
import globalConfig from '../config/globalConfig';
import {CLICommand} from '../command/cliCommand';
import {DecoratorFunction} from '../decorators/decoratorFunction';

const PRECONDITION_ERROR_CATEGORY = 'Authentication';
export enum AuthenticationType {
  OAuth = 'OAuth',
  ApiKey = 'ApiKey',
}

function getCurrentAuthenticationType() {
  const config = new Config(globalConfig.get().configDir).get();
  return config.anonymous
    ? AuthenticationType.ApiKey
    : AuthenticationType.OAuth;
}

export function IsAuthenticated(
  authenticationMethodAllowed: AuthenticationType[] = [
    AuthenticationType.ApiKey,
    AuthenticationType.OAuth,
  ]
): DecoratorFunction {
  return async function (_target: CLICommand) {
    const status = await getAuthenticationStatus();
    if (status === AuthenticationStatus.LOGGED_OUT) {
      throw new PreconditionError(
        'Not currently logged in. Run coveo auth:login first.',
        {category: PRECONDITION_ERROR_CATEGORY}
      );
    }

    if (status === AuthenticationStatus.EXPIRED) {
      throw new PreconditionError(
        'Authentication token is expired. Run coveo auth:login first.',
        {category: PRECONDITION_ERROR_CATEGORY}
      );
    }

    const currentAuthenticationType = getCurrentAuthenticationType();
    if (authenticationMethodAllowed.includes(currentAuthenticationType)) {
      return;
    }

    switch (currentAuthenticationType) {
      case 'ApiKey':
        throw new PreconditionError(
          'This command does not support API Key authentication. Please authenticate again using coveo auth:login',
          {category: PRECONDITION_ERROR_CATEGORY}
        );
      case 'OAuth':
        throw new PreconditionError(
          'This command does not support OAuth/web authentication. Please authenticate again using coveo auth:token',
          {category: PRECONDITION_ERROR_CATEGORY}
        );
    }
  };
}
