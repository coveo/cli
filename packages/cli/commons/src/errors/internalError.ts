import {CLIBaseError} from './cliBaseError';
import {dedent} from 'ts-dedent';

export class InternalError extends CLIBaseError {
  public name = 'Internal CLI Error';
  public constructor(message: string) {
    super();
    this.message = dedent`
      Internal CLI Error, please report to Coveo
      
      Message:
        ${message}`;
  }
}
