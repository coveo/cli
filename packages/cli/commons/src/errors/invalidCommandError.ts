import {CLIBaseError, SeverityLevel} from './cliBaseError';
import dedent from 'ts-dedent';

export class InvalidCommandError extends CLIBaseError {
  public name = 'Invalid CLI Command';
  public constructor(cause: string) {
    super('', {level: SeverityLevel.Error});
    this.message = dedent`
      An error has been found in the CLI, please report this error to Coveo.
      
      Cause:
      ${this.cause}

      Stack:
      ${this.stack}
      `;
  }
}
