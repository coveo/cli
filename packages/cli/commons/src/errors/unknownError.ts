import {CLIBaseError} from './cliBaseError';

export class UnknownError extends CLIBaseError {
  public name = 'Unknown CLI Error';
  public constructor(e?: unknown) {
    super(
      e instanceof Error
        ? {...e, name: `${UnknownError.name} - ${e.name}`}
        : undefined
    );
  }
}
