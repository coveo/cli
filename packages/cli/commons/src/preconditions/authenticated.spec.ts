jest.mock('../../platform/authenticatedClient');

import {fancyIt} from '@coveo/cli-commons-dev/lib/testUtils/it';
import {PreconditionError} from '../../errors/preconditionError';
import {
  AuthenticationStatus,
  getAuthenticationStatus,
} from '../../platform/authenticatedClient';
import {IsAuthenticated} from './authenticated';

describe('authenticated', () => {
  const mockedAuthenticatedClient = jest.mocked(getAuthenticationStatus);

  beforeEach(() => {
    jest.resetAllMocks();
    jest.fn();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe.each([
    [
      'is logged out',
      AuthenticationStatus.LOGGED_OUT,
      'Not currently logged in. Run coveo auth:login first.',
    ],
    [
      'authentication is expired',
      AuthenticationStatus.EXPIRED,
      'Authentication token is expired. Run coveo auth:login first.',
    ],
  ])(
    'when the user %s.',
    (
      _: unknown,
      authenticationStatus: AuthenticationStatus,
      expectedWarning: string
    ) => {
      fancyIt()(`warns '${expectedWarning}' and returns false`, async () => {
        mockedAuthenticatedClient.mockResolvedValue(authenticationStatus);
        await expect(IsAuthenticated()()).rejects.toThrow(PreconditionError);
      });
    }
  );

  describe('when the user is logged in', () => {
    fancyIt()('should not throw', async () => {
      mockedAuthenticatedClient.mockResolvedValue(
        AuthenticationStatus.LOGGED_IN
      );
      await expect(IsAuthenticated()()).resolves.not.toThrow();
    });
  });
});
