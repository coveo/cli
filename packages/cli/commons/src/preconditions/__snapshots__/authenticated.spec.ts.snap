// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticated when the user is logged in when the user is anonymous (meaning it logged in with an API Key) should throw when ApiKey is not allowed 1`] = `"This command does not support API Key authentication. Please authenticate again using coveo auth:login"`;

exports[`authenticated when the user is logged in when the user is not anonymous (meaning it logged in with an API Key) should throw when OAuth is not allowed 1`] = `"This command does not support OAuth/web authentication. Please authenticate again using coveo auth:token"`;
